crates/bevy_ecs/src/event/trigger.rs

```diff
@@ -0,0 +1,461 @@
+///   soundness misstep, as most [`Trigger`] implementations will invoke observers that are developed _for their specific [`Trigger`] type_.
+///   Without this constraint, something like [`GlobalTrigger`] could be called for _any_ [`Event`] type, even one that expects a different
+///   [`Trigger`] type. This would result in an unsound cast of [`GlobalTrigger`] reference.
+///   This is not expressed as an explicit type constraint,, as the `for<'a> Event::Trigger<'a>` lifetime can mismatch explicit lifetimes in
+///   some impls.
+/// - Read, understand, and abide by the lifetime constraints defined in the section below:
+///
+/// To understand why this must be unsafe, we must think carefully about the lifetimes involved.
+///
+/// The core challenge is that the lifetime of the `&mut E::Trigger<'_>` that we want to create
+/// within the [`ObserverRunner`] (which calls the [`Observer`](crate::observer::Observer)) may not be the same as the lifetime provided
+/// by [`Event::Trigger<'a>`].
+///
+/// If the lifetimes are the same, then we can safely create a `&mut E::Trigger<'_>` from the [`PtrMut`]
+/// passed to the observer runner function, and pass that to the observer system inside of ['On'](crate::observer::On).
+///
+/// If the lifetimes are not the same, then we must be careful to ensure that the `&mut E::Trigger<'_>` we create
+/// does not outlive the [`PtrMut`] that was passed to the observer runner function.
+/// Failing to do so could lead to use-after-free bugs.
+///
+/// This problem leaks into [`Trigger::trigger`] because the [`ObserverRunner`] function
+/// is called from within [`Trigger::trigger`], for most real-world [`Trigger`] implementations.
+///
+/// This is complex, and ways to simplify this would be welcome in the future!
+/// The safety requirements of this trait were prompted by this comment thread:
+/// <https://github.com/bevyengine/bevy/pull/20731#discussion_r2311907935>
+///
+/// which also discusses some alternative designs that were considered.
+///
+/// [`ObserverRunner`]: crate::observer::ObserverRunner
+pub unsafe trait Trigger<E: Event> {
+    /// Trigger the given `event`, running every [`Observer`](crate::observer::Observer) that matches the `event`, as defined by this
+    /// [`Trigger`] and the state stored on `self`.
+    ///
+    /// # Safety
+    /// - The [`CachedObservers`] `observers` must come from the [`DeferredWorld`] `world`
+    /// - [`TriggerContext`] must contain an [`EventKey`](crate::event::EventKey) that matches the `E` [`Event`] type
+    /// - `observers` must correspond to observers compatible with the event type `E`
+    /// - Read and abide by the "Safety" section defined in the top-level [`Trigger`] docs. Calling this function is
+    ///   unintuitively risky. _Do not use it directly unless you know what you are doing_. Importantly, this should only
+    ///   be called for an `event` whose [`Event::Trigger`] matches this trigger.
```

Comment:
This is a bit of a personal take but I would prefer the list of safety requirements to only contain actual conditions, and the more formal/mechanical they are to prove the better. Otherwise I fear that the safety comments when calling these methods become a kind of a pinky promise to not do UB, which IMO does not help actually verify the safety of them.

Reply:
Solid point. I'll adjust this!